trickle/trickle_server.go

```diff
@@ -0,0 +1,423 @@
+package trickle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+)
+
+// TODO sweep idle streams connections
+
+type StreamManager struct {
+	mutex   sync.RWMutex
+	streams map[string]*Stream
+}
+
+type Stream struct {
+	mutex       sync.RWMutex
+	segments    []*Segment
+	latestWrite int
+}
+
+type Segment struct {
+	idx    int
+	mutex  *sync.Mutex
+	cond   *sync.Cond
+	buffer *bytes.Buffer
+	closed bool
+}
+
+type SegmentSubscriber struct {
+	segment *Segment
+	readPos int
+}
+
+const maxSegmentsPerStream = 5
+
+const BaseServerPath = "/ai/live-video/"
+
+var FirstByteTimeout = errors.New("pending read timeout")
+
+func ConfigureServerWithMux(mux *http.ServeMux) {
+	/* TODO we probably want to configure the below
+	srv := &http.Server{
+		// say max segment size is 20 secs
+		// we can allow 2 * 20 secs given preconnects
+		ReadTimeout:  40 * time.Second,
+		WriteTimeout: 45 * time.Second,
+	}
+	*/
+
+	streamManager := &StreamManager{
+		streams: make(map[string]*Stream),
+	}
+	mux.HandleFunc("GET "+BaseServerPath+"{streamName}/{idx}", streamManager.handleGet)
+	mux.HandleFunc("POST "+BaseServerPath+"{streamName}/{idx}", streamManager.handlePost)
+	mux.HandleFunc("DELETE "+BaseServerPath+"{streamName}", streamManager.handleDelete)
+}
+
+func (sm *StreamManager) getStream(streamName string) (*Stream, bool) {
```

Comment:
Nit: Don't you want to reorder functions to follow the [stepdown rule](https://medium.com/@alankarrocks/enhancing-code-readability-with-the-step-down-rule-3185ff78650b). I guess it's simpler to read the code.

Reply:
Old habits from C, where you have to define (or at least declare) functions before using them. I am somewhat used to looking _up_ for a given definition, rather than down but I suppose it does not really matter in golang. If the convention is step down style, we can do that but there will always be some subjectivity about what is "important" enough to come first (declaration-first at least gives something of a rule to follow)